mopro-core/src/middleware/gpu_explorations/metal/msm.rs

```diff
@@ -0,0 +1,359 @@
+    // convert window_starts to u32 to give the exact storage need for GPU
+    let window_starts_buffer = state.alloc_buffer_data(
+        &(window_starts
+            .iter()
+            .map(|x| *x as u32)
+            .collect::<Vec<u32>>()),
+    );
+
+    // do msm accumulation in windows-wise fashion all in MSL
+    let calc_bucket_pipe = state.setup_pipeline("accumulation_in_window_wise").unwrap();
+    autoreleasepool(|| {
+        let (command_buffer, command_encoder) = state.setup_command(
+            &calc_bucket_pipe,
+            Some(&[
+                (0, &window_size_buffer),
+                (1, &instances_size_buffer),
+                (2, &window_starts_buffer),
+                (3, &scalar_buffer),
+                (4, &base_buffer),
+                (5, &buckets_matrix_buffer),
+            ]),
+        );
+        command_encoder.dispatch_thread_groups(
+            MTLSize::new(1, 1, 1),
+            MTLSize::new(window_starts.len() as u64, 1, 1),
+        );
+        command_encoder.end_encoding();
+        command_buffer.commit();
+        command_buffer.wait_until_completed();
+    });
+
+    // retrieve and parse the result from GPU
+    let buckets_matrix = {
+        let raw_limbs = MetalState::retrieve_contents::<u32>(&buckets_matrix_buffer);
+        cfg_into_iter!(raw_limbs)
+            .chunks(24)
+            .map(|chunk| {
+                G::new_unchecked(
+                    Fq::from_u32_limbs(&chunk[0..8]),
+                    Fq::from_u32_limbs(&chunk[8..16]),
+                    Fq::from_u32_limbs(&chunk[16..24]),
+                )
+            })
+            .collect::<Vec<_>>()
+    };
+
+    // include the last windox idx
+    let bucket_starts: Vec<usize> = (0..buckets_matrix.len() + buckets_size)
+        .step_by(buckets_size)
+        .collect();
+    let window_sums: Vec<_> = cfg_into_iter!(window_starts.clone())
+        .enumerate()
+        .map(|(idx, w_start)| {
+            // only process unit scalars once in the first window.
+            let mut res = zero;
+            if w_start == 0 {
+                for i in 0..instances_size {
+                    if scalars[i] == one {
+                        res += points[i];
+                    }
+                }
+            }
+
+            let buckets = buckets_matrix[bucket_starts[idx]..bucket_starts[idx + 1]].to_vec();
+            let mut running_sum = zero;
+            buckets.into_iter().rev().for_each(|b| {
+                running_sum += &b;
+                res += &running_sum;
+            });
+            res
+        })
+        .collect();
+
+    let lowest = *window_sums.first().unwrap();
+
+    Ok(lowest
+        + &window_sums[1..]
+            .iter()
+            .rev()
+            .fold(zero, |mut total, sum_i| {
+                total += sum_i;
+                for _ in 0..window_size {
+                    total = total.double();
+                }
+                total
+            }))
+}
```

Comment:
Possible optimization direction: move all of this msm computation to GPU using metal